add publish, save and profile code into toolbars

javierluraschi opened the pull request:
Profiler improvements:
- Add additional command entries into profiler toolbar: Publish and save-as-profile
- Save unnamed files as Profile# instead of Untitled#
- Use dynamic commands for save-profile-as command

@jjallaire the "profile-code" icon was crafted from the "run-code" and "profile" commands. Could we get Pauls feedback and iterate as needed?

<img width="1440" alt="screen shot 2016-03-04 at 11 16 43 am" src="https://cloud.githubusercontent.com/assets/3478847/13537710/156fcbc2-e1fb-11e5-9df5-6364bd756dc3.png">

<img width="1440" alt="screen shot 2016-03-04 at 11 16 51 am" src="https://cloud.githubusercontent.com/assets/3478847/13537713/198a43d6-e1fb-11e5-931a-72ee1f3bc1b9.png">

jmcphers commented on src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/model/ProfilerContents.java:
```diff
@@ -45,4 +45,12 @@ public final native boolean getCreateProfile() /*-{
+      return this.htmlLocalPath;
+   }-*/;
+   
+   public final native boolean isUserSaved() /*-{
```
Is this used?

javierluraschi replied:
Sort of, fix was not complete, now it is. This new property tracks the saved state of the document to allow users to see the "save" command enabled only once while their doc has never ben saved. Once it's saved, it can only ve saved again with the "save as" command. This matches the behavior of other docs in the IDE.

jjallaire replied:
I don't have a problem with shuffling around shortcuts (we've historically
done quite a bit of that, especially for less frequently used commands).

One final tweak: I'd eliminate the icon for that command so that it's
display is consistent with the rest of the items on the menu.


On Fri, Mar 4, 2016 at 5:16 PM, Javier Luraschi <notifications@github.com>
wrote:

> In
> src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/model/ProfilerContents.java
> <https://github.com/rstudio/rstudio/pull/619#discussion_r55097221>:
>
> > @@ -45,4 +45,12 @@ public final native boolean getCreateProfile() /*-{
> >     public final native String getName() /*-{
> >        return this.name;
> >     }-*/;
> > +
> > +   public final native String getHtmlLocalPath() /*-{
> > +      return this.htmlLocalPath;
> > +   }-*/;
> > +
> > +   public final native boolean isUserSaved() /*-{
>
> Sort of, fix was not complete, now it is. This new property tracks the
> saved state of the document to allow users to see the "save" command
> enabled only once while their doc has never ben saved. Once it's saved, it
> can only ve saved again with the "save as" command. This matches the
> behavior of other docs in the IDE.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/rstudio/pull/619/files#r55097221>.
>

jjallaire commented:
LGTW!

A couple comments about the profiling toolbar icon:

1. I think it should go on the "Code Wizard" menu rather than haven't it's
own toolbar button.

2. Your derivative icon is fine but I also think we could just use the
"Profile" without the source document overlay (could be less busy and more
clear).



On Fri, Mar 4, 2016 at 2:23 PM, Javier Luraschi <notifications@github.com>
wrote:

> Profiler improvements:
>
>    - Add additional command entries into profiler toolbar: Publish and
>    save-as-profile
>    - Save unnamed files as Profile# instead of Untitled#
>    - Use dynamic commands for save-profile-as command
>
> @jjallaire <https://github.com/jjallaire> the "profile-code" icon was
> crafted from the "run-code" and "profile" commands. Could we get Pauls
> feedback and iterate as needed?
>
> [image: screen shot 2016-03-04 at 11 16 43 am]
> <https://cloud.githubusercontent.com/assets/3478847/13537710/156fcbc2-e1fb-11e5-9df5-6364bd756dc3.png>
>
> [image: screen shot 2016-03-04 at 11 16 51 am]
> <https://cloud.githubusercontent.com/assets/3478847/13537713/198a43d6-e1fb-11e5-931a-72ee1f3bc1b9.png>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/rstudio/rstudio/pull/619
> Commit Summary
>
>    - add toolbar command to publish from profiler view
>    - add support to name untitled documents based on edittarget
>    requirements
>    - add support to save as profile from toolbar
>    - add profile code button to source toolbar
>    - use dynamic command for save profile as
>
> File Changes
>
>    - *M* src/cpp/session/modules/SessionProfiler.R
>    <https://github.com/rstudio/rstudio/pull/619/files#diff-0> (5)
>    - *M*
>    src/gwt/src/org/rstudio/studio/client/common/filetypes/ProfilerType.java
>    <https://github.com/rstudio/rstudio/pull/619/files#diff-1> (1)
>    - *M*
>    src/gwt/src/org/rstudio/studio/client/workbench/commands/Commands.cmd.xml
>    <https://github.com/rstudio/rstudio/pull/619/files#diff-2> (3)
>    - *A*
>    src/gwt/src/org/rstudio/studio/client/workbench/commands/profileCode.png
>    <https://github.com/rstudio/rstudio/pull/619/files#diff-3> (0)
>    - *A*
>    src/gwt/src/org/rstudio/studio/client/workbench/commands/saveProfileAs.png
>    <https://github.com/rstudio/rstudio/pull/619/files#diff-4> (0)
>    - *M*
>    src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java
>    <https://github.com/rstudio/rstudio/pull/619/files#diff-5> (19)
>    - *M*
>    src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/EditingTarget.java
>    <https://github.com/rstudio/rstudio/pull/619/files#diff-6> (2)
>    - *M*
>    src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/EditingTargetSource.java
>    <https://github.com/rstudio/rstudio/pull/619/files#diff-7> (29)
>    - *M*
>    src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/codebrowser/CodeBrowserEditingTarget.java
>    <https://github.com/rstudio/rstudio/pull/619/files#diff-8> (6)
>    - *M*
>    src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/ProfilerEditingTarget.java
>    <https://github.com/rstudio/rstudio/pull/619/files#diff-9> (58)
>    - *M*
>    src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/ProfilerEditingTargetWidget.java
>    <https://github.com/rstudio/rstudio/pull/619/files#diff-10> (21)
>    - *M*
>    src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/model/ProfileOperationResponse.java
>    <https://github.com/rstudio/rstudio/pull/619/files#diff-11> (4)
>    - *M*
>    src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/profiler/model/ProfilerContents.java
>    <https://github.com/rstudio/rstudio/pull/619/files#diff-12> (8)
>    - *M*
>    src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/TextEditingTarget.java
>    <https://github.com/rstudio/rstudio/pull/619/files#diff-13> (5)
>    - *M*
>    src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/TextEditingTargetWidget.java
>    <https://github.com/rstudio/rstudio/pull/619/files#diff-14> (2)
>    - *M*
>    src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/urlcontent/UrlContentEditingTarget.java
>    <https://github.com/rstudio/rstudio/pull/619/files#diff-15> (5)
>
> Patch Links:
>
>    - https://github.com/rstudio/rstudio/pull/619.patch
>    - https://github.com/rstudio/rstudio/pull/619.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/rstudio/pull/619>.
>

javierluraschi commented:
@jjallaire OK!

1. Moved.
2. Changed back to standard profile icon and command to code wizard.

BTW. @jjallaire @jmcphers @kevinushey: I just noticed that I chose a keyboard shortcut that is already in use by "Show Diagnostics (Project)". Is it OK if I renamed the diagnostics shortcut from `Cmd+Shift+Alt+P` to `Cmd+Shift+Alt+D`? I understand that some people might reference their command in their tutorials or books, however, this seems unlikely since it does not seem like a high-profile command. Thoughts?

<img width="1440" alt="screen shot 2016-03-04 at 2 24 39 pm" src="https://cloud.githubusercontent.com/assets/3478847/13542060/3839f168-e215-11e5-82d6-52381ac41d95.png">

kevinushey commented:
I agree with shuffling around the shortcuts like this is as well -- `Cmd+Shift+Alt+P` is relatively accessible and it's likely that the profiler will get a lot of use, whereas diagnostics mostly just run in the background (without requiring explicit input from the user)
